### DISCUSSION TITLE
Two Questions

### DISCUSSION THREAD

**Discussion by angeltxy** [ohler55/ojg#177] [Q&A] [answered]
1. Does filter expression supports regex, just like the example below?
`{
    "store": {
        "book": [
            {
                "category": "reference",
                "author": "Nigel Rees",
                "title": "Sayings of the Century",
                "price": 8.95
            },
            {
                "category": "fiction",
                "author": "Evelyn Waugh",
                "title": "Sword of Honour",
                "price": 12.99
            },
            {
                "category": "fiction",
                "author": "Herman Melville",
                "title": "Moby Dick",
                "isbn": "0-553-21311-3",
                "price": 8.99
            },
            {
                "category": "fiction",
                "author": "J. R. R. Tolkien",
                "title": "The Lord of the Rings",
                "isbn": "0-395-19395-8",
                "price": 22.99
            }
        ],
        "bicycle": {
            "color": "red",
            "price": 19.95
        }
    },
    "expensive": 10
}`

JsonPath: $..book[?(@.author =~ /.*REES/i)]
Result: [{"category": "reference", "author": "Nigel Rees", "title": "Sayings of the Century", "price": 8.95}]

2. Do you think about support functions like this below:
![image](https://github.com/ohler55/ojg/assets/22721819/4755270b-85b6-41ea-ab87-f7be6e845f80)
especially function keys.  The example:
json: 
`{
      "111": "aaa",
      "222": "bbb",
      "333": {
            "333aaa": "333_aaa",
	    "333bbb": "333_bbb",
	    "333ccc": {
				"xxx": "hahaha",
				"yyy": "wuwuwu",
				"zzz": "kekeke"
			}
		}
	}`

JsonPath: $.333.333ccc.keys()[0:1]
Result: ["xxx"]
In our program, to get the function of keys, we split the jsonpath with ".keys()" to get sub jsonpath slice, and the output of  previous sub jsonpath in the slice to be used as the input of the next, code like this:
`
           ```
 func JsonPathRead(data any, path string) any {
	      ps := strings.Split(path, ".keys()")
	      newPaths := make([]string, 0)
	      for i, p := range ps {
		      if i == 0 {
			      newPaths = append(newPaths, p)
		      } else {
			      newPaths = append(newPaths, fmt.Sprintf("$%s", p))
		      }
		      if i < len(ps)-1 {
			      newPaths = append(newPaths, "$.keys()")
		      }
	      }
	      var output any
	      firstRead := false
	      for _, p := range newPaths {
		      if p == "$" {
			      continue
		      }
		      if !firstRead {
			      output = read(data, p)
			      firstRead = true
		      } else {
			      output = read(output, p)
		      }
	      }
	      return output
      }
```
            
      func read(data any, path string) any {
            if path == "$.keys()" {
                  if IsMap(data) {
                        if m, ok := data.(map[string]any); ok {
                              r := make([]any, 0)
			     for k := range m {
                                    r = append(r, k)
                              }
                              return r
                        }
                  }
            }
            x, err := jp.ParseString(path)
            if err != nil {
                  return nil
            }
            results := x.Get(data)
            if len(results) == 1 {
                  return results[0]
            }
            return results
      }
`

Best regards

**Comment by ohler55** [comment, marked as answer]
As an answer to the first question, yes OjG does support regex. The correct syntax for the JSONPath is `$..book[?(@.author ~= /(?i).*REES/)]`.

**Reply by angeltxy** [reply]
I got it, thank you😄

**Comment by ohler55** [comment]
In answer to question 2 please take a look at the JSONPath spec at https://datatracker.ietf.org/doc/rfc9535. OjG supports the length, count, search, and match functions. The syntax is a bit different than what is proposed though. OjG attempts to follow the spec when providing specific features such as the use of functions like length. Instead of `@.data.length()` the use would be `length(@.data)`.

Although there is currently not keys function in OjG, it is possible to obtain keys using the jp.Walk function and then doing some filtering for the paths of interest.

**Reply by angeltxy** [reply]
Thanks again🙏

**Comment by ohler55** [comment]
I was thinking of allowing users to add their own functions but haven't worked that out yet. Use would be like length, count, etc.

**Reply by angeltxy** [reply]
That's great, looking forward to it very much👍👍👍

**Reply by ohler55** [reply]
Release v1.23.0 has the ability to add your own custom functions.

**Reply by angeltxy** [reply]
Wow great👍, I am very surprised and grateful😄, I'll try this new feature later, and thanks again😄
